runners/samza/src/main/java/org/apache/beam/runners/samza/translation/PortableTranslationContext.java

```diff
@@ -0,0 +1,176 @@
+    this.streamGraph = streamGraph;
+    this.options = options;
+  }
+
+  public SamzaPipelineOptions getSamzaPipelineOptions() {
+    return this.options;
+  }
+
+  public void setCurrentTopologicalId(int id) {
+    this.topologicalId = id;
+  }
+
+  public int getCurrentTopologicalId() {
+    return this.topologicalId;
+  }
+
+  public <T> List<MessageStream<OpMessage<T>>> getAllInputMessageStreams(
+      PipelineNode.PTransformNode transform) {
+    final Collection<String> inputStreamIds = transform.getTransform().getInputsMap().values();
+    return inputStreamIds.stream().map(this::<T>getMessageStreamById).collect(Collectors.toList());
+  }
+
+  public <T> MessageStream<OpMessage<T>> getOneInputMessageStream(
+      PipelineNode.PTransformNode transform) {
+    String id = Iterables.getOnlyElement(transform.getTransform().getInputsMap().values());
+    return getMessageStreamById(id);
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T> MessageStream<OpMessage<T>> getMessageStreamById(String id) {
+    return (MessageStream<OpMessage<T>>) messsageStreams.get(id);
+  }
+
+  public String getInputId(PipelineNode.PTransformNode transform) {
+    return Iterables.getOnlyElement(transform.getTransform().getInputsMap().values());
+  }
+
+  public String getOutputId(PipelineNode.PTransformNode transform) {
+    return Iterables.getOnlyElement(transform.getTransform().getOutputsMap().values());
+  }
+
+  public <T> void registerMessageStream(String id, MessageStream<OpMessage<T>> stream) {
+    if (messsageStreams.containsKey(id)) {
+      throw new IllegalArgumentException("Stream already registered for id: " + id);
+    }
+    messsageStreams.put(id, stream);
+  }
+
+  /** Register an input stream, using the PCollection id as the config id. */
+  public void registerInputMessageStream(String id) {
+    registerInputMessageStreamWithStreamId(id, id);
+  }
+
+  /** Get output stream by stream id. */
+  public <T> OutputStream<T> getOutputStreamById(String outputStreamId) {
+    return streamGraph.getOutputStream(outputStreamId);
+  }
+
+  /**
+   * Register an input stream with certain config id.
+   *
+   * @param id id of the PCollection in the input/output of PTransform
+   * @param streamId samza stream id which user can use to customize the stream level config
+   */
+  public <T> void registerInputMessageStreamWithStreamId(String id, String streamId) {
+    // we want to register it with the Samza graph only once per i/o stream
+    if (registeredInputStreams.contains(streamId)) {
+      return;
+    }
+    final MessageStream<OpMessage<T>> stream =
+        streamGraph
+            .<org.apache.samza.operators.KV<?, OpMessage<T>>>getInputStream(streamId)
+            .map(org.apache.samza.operators.KV::getValue);
+
+    registerMessageStream(id, stream);
+    registeredInputStreams.add(streamId);
+  }
+
+  public WindowedValue.WindowedValueCoder instantiateCoder(
```

Comment:
Seems instantiateCoder() and getPortableWindowStrategy() can be static util classes in the util package, or inline it if there is only one use case.